File: apps/android/app/src/main/kotlin/com/epicstartup/tenant/core/theme/RgbaColor.kt
package com.epicstartup.tenant.core.theme

/**
 * A plain sRGB color with alpha in `0...1`.
 *
 * The core deliberately does not import `android.graphics`: the same theme math
 * is used by the app and by the JVM unit tests. [toArgb] is the only bridge.
 */
class RgbaColor private constructor(
	val red: Double,
	val green: Double,
	val blue: Double,
	val alpha: Double,
) {
	fun withAlpha(alpha: Double): RgbaColor = of(red, green, blue, alpha)

	fun mixedWith(other: RgbaColor, amount: Double): RgbaColor {
		val ratio = amount.coerceIn(0.0, 1.0)
		return of(
			red + (other.red - red) * ratio,
			green + (other.green - green) * ratio,
			blue + (other.blue - blue) * ratio,
			alpha + (other.alpha - alpha) * ratio,
		)
	}

	/** WCAG relative luminance — used to pick readable foreground text. */
	val relativeLuminance: Double
		get() {
			fun channel(value: Double): Double =
				if (value <= 0.03928) value / 12.92 else Math.pow((value + 0.055) / 1.055, 2.4)
			return 0.2126 * channel(red) + 0.7152 * channel(green) + 0.0722 * channel(blue)
		}

	val isLight: Boolean
		get() = relativeLuminance > 0.5

	fun contrastRatio(against: RgbaColor): Double {
		val first = maxOf(relativeLuminance, against.relativeLuminance)
		val second = minOf(relativeLuminance, against.relativeLuminance)
		return (first + 0.05) / (second + 0.05)
	}

	val hexString: String
		get() {
			fun hex(value: Double): String {
				val scaled = Math.round(value * 255).toInt().coerceIn(0, 255)
				return scaled.toString(16).padStart(2, '0')
			}
			return "#${hex(red)}${hex(green)}${hex(blue)}"
		}

	/** `0xAARRGGBB`, ready for `android.graphics.Color` consumers. */
	fun toArgb(): Int {
		val a = Math.round(alpha * 255).toInt().coerceIn(0, 255)
		val r = Math.round(red * 255).toInt().coerceIn(0, 255)
		val g = Math.round(green * 255).toInt().coerceIn(0, 255)
		val b = Math.round(blue * 255).toInt().coerceIn(0, 255)
		return (a shl 24) or (r shl 16) or (g shl 8) or b
	}

	override fun equals(other: Any?): Boolean =
		other is RgbaColor &&
			other.red == red &&
			other.green == green &&
			other.blue == blue &&
			other.alpha == alpha

	override fun hashCode(): Int {
		var result = red.hashCode()
		result = 31 * result + green.hashCode()
		result = 31 * result + blue.hashCode()
		result = 31 * result + alpha.hashCode()
		return result
	}

	override fun toString(): String = "RgbaColor($hexString, alpha=$alpha)"

	companion object {
		val black = of(0.0, 0.0, 0.0)
		val white = of(1.0, 1.0, 1.0)

		fun of(red: Double, green: Double, blue: Double, alpha: Double = 1.0): RgbaColor =
			RgbaColor(component(red), component(green), component(blue), component(alpha))

		/** `coerceIn` lets NaN through, so sanitise before clamping. */
		private fun component(value: Double): Double =
			if (!value.isFinite()) 0.0 else value.coerceIn(0.0, 1.0)

		val named: Map<String, RgbaColor> = mapOf(
			"transparent" to of(0.0, 0.0, 0.0, 0.0),
			"white" to white,
			"black" to black,
			"red" to of(1.0, 0.0, 0.0),
			"green" to of(0.0, 0.5019607843, 0.0),
			"blue" to of(0.0, 0.0, 1.0),
			"gray" to of(0.5019607843, 0.5019607843, 0.5019607843),
			"grey" to of(0.5019607843, 0.5019607843, 0.5019607843),
			"silver" to of(0.7529411765, 0.7529411765, 0.7529411765),
			"orange" to of(1.0, 0.6470588235, 0.0),
			"purple" to of(0.5019607843, 0.0, 0.5019607843),
			"yellow" to of(1.0, 1.0, 0.0),
		)

		// MARK: CSS parsing

		/**
		 * Parses the color formats the org theme compiler emits.
		 *
		 * shadcn themes are authored in `oklch()`, but orgs can override
		 * individual tokens with hex/rgb/hsl values, so all four are supported.
		 * Anything the parser cannot understand (for example CSS relative color
		 * syntax such as `oklch(from var(--destructive) min(l, 0.42) c h)`)
		 * returns `null` and the caller keeps its fallback token.
		 */
		fun from(cssValue: String): RgbaColor? {
			val value = cssValue.trim().lowercase()
			if (value.isEmpty()) return null
			return when {
				value.startsWith("#") -> fromHex(value)
				value.startsWith("rgb") -> fromFunctional(value, "rgb")
				value.startsWith("hsl") -> fromHsl(value)
				value.startsWith("oklch(") -> fromOklch(value)
				value.startsWith("oklab(") -> fromOklab(value)
				else -> named[value]
			}
		}

		private fun fromHex(value: String): RgbaColor? {
			val digits = value.drop(1)
			if (digits.isEmpty()) return null
			if (!digits.all { it.isDigit() || it in 'a'..'f' }) return null

			fun component(text: String): Double {
				val parsed = text.toIntOrNull(16) ?: return 0.0
				return parsed / 255.0
			}

			return when (digits.length) {
				3, 4 -> {
					val characters = digits.map { it.toString() }
					of(
						component(characters[0] + characters[0]),
						component(characters[1] + characters[1]),
						component(characters[2] + characters[2]),
						if (digits.length == 4) component(characters[3] + characters[3]) else 1.0,
					)
				}
				6, 8 -> {
					val characters = digits.map { it.toString() }
					of(
						component(characters[0] + characters[1]),
						component(characters[2] + characters[3]),
						component(characters[4] + characters[5]),
						if (digits.length == 8) component(characters[6] + characters[7]) else 1.0,
					)
				}
				else -> null
			}
		}

		/** Splits `rgb(1 2 3 / 40%)` or `rgb(1, 2, 3, 0.4)` into components. */
		private fun splitFunction(value: String, prefix: String): Pair<List<String>, String?>? {
			if (!value.startsWith(prefix)) return null
			val open = value.indexOf('(')
			if (open < 0 || !value.endsWith(")")) return null
			val inner = value.substring(open + 1, value.length - 1)
			val parts = inner.replace(',', ' ').split('/')
			if (parts.isEmpty()) return null
			val alpha = if (parts.size > 1) parts[1].trim() else null
			val components = parts[0].split(' ', '\t').filter { it.isNotEmpty() }
			return components to alpha
		}

		private fun componentValue(text: String, scale: Double): Double? {
			val value = if (text.endsWith("%")) {
				text.dropLast(1).toDoubleOrNull()?.div(100)
			} else {
				text.toDoubleOrNull()?.div(scale)
			}
			return value?.takeIf { it.isFinite() }
		}

		/**
		 * `null` when an explicit alpha is present but unparsable, so the caller
		 * can keep its fallback token instead of rendering an opaque color.
		 */
		private fun alphaValue(text: String?): Double? {
			if (text.isNullOrEmpty()) return 1.0
			val value = if (text.endsWith("%")) {
				text.dropLast(1).toDoubleOrNull()?.div(100)
			} else {
				text.toDoubleOrNull()
			}
			return value?.takeIf { it.isFinite() }
		}

		private fun fromFunctional(value: String, prefix: String): RgbaColor? {
			val (components, alpha) = splitFunction(value, prefix) ?: return null
			if (components.size < 3) return null
			val red = componentValue(components[0], 255.0) ?: return null
			val green = componentValue(components[1], 255.0) ?: return null
			val blue = componentValue(components[2], 255.0) ?: return null
			val resolvedAlpha = alpha ?: components.getOrNull(3)
			val alphaValue = alphaValue(resolvedAlpha) ?: return null
			return of(red, green, blue, alphaValue)
		}

		private fun fromHsl(value: String): RgbaColor? {
			val (components, alpha) = splitFunction(value, "hsl") ?: return null
			if (components.size < 3) return null
			val hue = components[0].replace("deg", "").toDoubleOrNull() ?: return null
			val saturation = componentValue(components[1], 100.0) ?: return null
			val lightness = componentValue(components[2], 100.0) ?: return null
			val alphaValue = alphaValue(alpha ?: components.getOrNull(3)) ?: return null
			return fromHslComponents(hue, saturation, lightness, alphaValue)
		}

		fun fromHslComponents(
			hue: Double,
			saturation: Double,
			lightness: Double,
			alpha: Double = 1.0,
		): RgbaColor {
			val hueDegrees = hue % 360
			val normalizedHue = if (hueDegrees < 0) hueDegrees + 360 else hueDegrees
			val chroma = (1 - Math.abs(2 * lightness - 1)) * saturation
			val secondary = chroma * (1 - Math.abs((normalizedHue / 60) % 2 - 1))
			val match = lightness - chroma / 2

			val rgb = when {
				normalizedHue < 60 -> Triple(chroma, secondary, 0.0)
				normalizedHue < 120 -> Triple(secondary, chroma, 0.0)
				normalizedHue < 180 -> Triple(0.0, chroma, secondary)
				normalizedHue < 240 -> Triple(0.0, secondary, chroma)
				normalizedHue < 300 -> Triple(secondary, 0.0, chroma)
				else -> Triple(chroma, 0.0, secondary)
			}

			return of(rgb.first + match, rgb.second + match, rgb.third + match, alpha)
		}

		/** `oklch(0.205 0 0)` / `oklch(62.3% 0.214 259.815)` / `oklch(1 0 0 / 50%)`. */
		private fun fromOklch(value: String): RgbaColor? {
			if (value.contains("from ")) return null
			val (components, alpha) = splitFunction(value, "oklch") ?: return null
			if (components.size < 3) return null
			val lightness = componentValue(components[0], 1.0) ?: return null
			val chroma = components[1].toDoubleOrNull()?.takeIf { it.isFinite() } ?: return null
			val alphaValue = alphaValue(alpha) ?: return null
			val hueText = components[2].replace("deg", "")
			val hue = when {
				hueText == "none" -> 0.0
				// An unparsable hue would silently become red; let the caller fall back.
				else -> hueText.toDoubleOrNull()?.takeIf { it.isFinite() } ?: return null
			}
			return fromOklchComponents(lightness, chroma, hue, alphaValue)
		}

		private fun fromOklab(value: String): RgbaColor? {
			val (components, alpha) = splitFunction(value, "oklab") ?: return null
			if (components.size < 3) return null
			val lightness = componentValue(components[0], 1.0) ?: return null
			val a = components[1].toDoubleOrNull()?.takeIf { it.isFinite() } ?: return null
			val b = components[2].toDoubleOrNull()?.takeIf { it.isFinite() } ?: return null
			val alphaValue = alphaValue(alpha) ?: return null
			return fromOklabComponents(lightness, a, b, alphaValue)
		}

		/** OKLCH → OKLab → linear sRGB → sRGB (Björn Ottosson's matrices). */
		fun fromOklchComponents(
			lightness: Double,
			chroma: Double,
			hue: Double,
			alpha: Double = 1.0,
		): RgbaColor {
			val hueRadians = hue * Math.PI / 180
			return fromOklabComponents(
				lightness = lightness,
				a = chroma * Math.cos(hueRadians),
				b = chroma * Math.sin(hueRadians),
				alpha = alpha,
			)
		}

		fun fromOklabComponents(
			lightness: Double,
			a: Double,
			b: Double,
			alpha: Double = 1.0,
		): RgbaColor {
			val long = Math.pow(lightness + 0.3963377774 * a + 0.2158037573 * b, 3.0)
			val medium = Math.pow(lightness - 0.1055613458 * a - 0.0638541728 * b, 3.0)
			val short = Math.pow(lightness - 0.0894841775 * a - 1.2914855480 * b, 3.0)

			val linearRed = 4.0767416621 * long - 3.3077115913 * medium + 0.2309699292 * short
			val linearGreen = -1.2684380046 * long + 2.6097574011 * medium - 0.3413193965 * short
			val linearBlue = -0.0041960863 * long - 0.7034186147 * medium + 1.7076147010 * short

			fun gamma(value: Double): Double =
				if (value <= 0.0031308) 12.92 * value else 1.055 * Math.pow(maxOf(value, 0.0), 1 / 2.4) - 0.055

			return of(
				gamma(linearRed),
				gamma(linearGreen),
				gamma(linearBlue),
				alpha,
			)
		}
	}
}

File: apps/ios/Sources/TenantKit/Theme/RGBAColor.swift
import Foundation

/// A plain sRGB color with alpha in `0...1`.
///
/// TenantKit deliberately does not import SwiftUI: the same theme math is used
/// by the iOS app and by the Linux test suite.
public struct RGBAColor: Equatable, Sendable, Hashable {
	public let red: Double
	public let green: Double
	public let blue: Double
	public let alpha: Double

	public init(red: Double, green: Double, blue: Double, alpha: Double = 1) {
		self.red = min(max(red, 0), 1)
		self.green = min(max(green, 0), 1)
		self.blue = min(max(blue, 0), 1)
		self.alpha = min(max(alpha, 0), 1)
	}

	public static let black = RGBAColor(red: 0, green: 0, blue: 0)
	public static let white = RGBAColor(red: 1, green: 1, blue: 1)

	public static let named: [String: RGBAColor] = [
		"transparent": RGBAColor(red: 0, green: 0, blue: 0, alpha: 0),
		"white": .white,
		"black": .black,
		"red": RGBAColor(red: 1, green: 0, blue: 0),
		"green": RGBAColor(red: 0, green: 0.5019607843, blue: 0),
		"blue": RGBAColor(red: 0, green: 0, blue: 1),
		"gray": RGBAColor(red: 0.5019607843, green: 0.5019607843, blue: 0.5019607843),
		"grey": RGBAColor(red: 0.5019607843, green: 0.5019607843, blue: 0.5019607843),
		"silver": RGBAColor(red: 0.7529411765, green: 0.7529411765, blue: 0.7529411765),
		"orange": RGBAColor(red: 1, green: 0.6470588235, blue: 0),
		"purple": RGBAColor(red: 0.5019607843, green: 0, blue: 0.5019607843),
		"yellow": RGBAColor(red: 1, green: 1, blue: 0),
	]

	public func withAlpha(_ alpha: Double) -> RGBAColor {
		RGBAColor(red: red, green: green, blue: blue, alpha: alpha)
	}

	public func mixed(with other: RGBAColor, amount: Double) -> RGBAColor {
		let ratio = min(max(amount, 0), 1)
		return RGBAColor(
			red: red + (other.red - red) * ratio,
			green: green + (other.green - green) * ratio,
			blue: blue + (other.blue - blue) * ratio,
			alpha: alpha + (other.alpha - alpha) * ratio
		)
	}

	/// WCAG relative luminance — used to pick readable foreground text when an
	/// org theme leaves a token missing.
	public var relativeLuminance: Double {
		func channel(_ value: Double) -> Double {
			value <= 0.03928 ? value / 12.92 : pow((value + 0.055) / 1.055, 2.4)
		}
		return 0.2126 * channel(red) + 0.7152 * channel(green) + 0.0722 * channel(blue)
	}

	public var isLight: Bool {
		relativeLuminance > 0.5
	}

	public func contrastRatio(against other: RGBAColor) -> Double {
		let first = max(relativeLuminance, other.relativeLuminance)
		let second = min(relativeLuminance, other.relativeLuminance)
		return (first + 0.05) / (second + 0.05)
	}

	public var hexString: String {
		func hex(_ value: Double) -> String {
			let scaled = Int((value * 255).rounded())
			return String(format: "%02x", min(max(scaled, 0), 255))
		}
		return "#\(hex(red))\(hex(green))\(hex(blue))"
	}

	// MARK: - CSS parsing

	/// Parses the color formats the org theme compiler emits.
	///
	/// shadcn themes are authored in `oklch()`, but orgs can override individual
	/// tokens with hex/rgb/hsl values, so all four are supported. Anything the
	/// parser cannot understand (for example CSS relative color syntax such as
	/// `oklch(from var(--destructive) min(l, 0.42) c h)`) returns `nil` and the
	/// caller keeps its fallback token.
	public static func from(cssValue: String) -> RGBAColor? {
		let value = cssValue.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
		guard !value.isEmpty else { return nil }

		if value.hasPrefix("#") {
			return fromHex(value)
		}
		if value.hasPrefix("rgb") {
			return fromFunctional(value, prefix: "rgb")
		}
		if value.hasPrefix("hsl") {
			return fromHSL(value)
		}
		if value.hasPrefix("oklch(") {
			return fromOKLCH(value)
		}
		if value.hasPrefix("oklab(") {
			return fromOKLab(value)
		}
		return named[value]
	}

	private static func fromHex(_ value: String) -> RGBAColor? {
		let digits = String(value.dropFirst())
		guard digits.allSatisfy({ $0.isHexDigit }) else { return nil }

		func component(_ text: String) -> Double {
			guard let intValue = Int(text, radix: 16) else { return 0 }
			return Double(intValue) / 255
		}

		switch digits.count {
		case 3, 4:
			let characters = Array(digits).map(String.init)
			return RGBAColor(
				red: component(characters[0] + characters[0]),
				green: component(characters[1] + characters[1]),
				blue: component(characters[2] + characters[2]),
				alpha: digits.count == 4 ? component(characters[3] + characters[3]) : 1
			)
		case 6, 8:
			let characters = Array(digits).map(String.init)
			return RGBAColor(
				red: component(characters[0] + characters[1]),
				green: component(characters[2] + characters[3]),
				blue: component(characters[4] + characters[5]),
				alpha: digits.count == 8 ? component(characters[6] + characters[7]) : 1
			)
		default:
			return nil
		}
	}

	/// Splits `rgb(1 2 3 / 40%)` or `rgb(1, 2, 3, 0.4)` into components.
	private static func splitFunction(_ value: String, prefix: String) -> (components: [String], alpha: String?)? {
		guard let open = value.firstIndex(of: "("), value.hasSuffix(")") else { return nil }
		guard value.hasPrefix(prefix) else { return nil }
		let inner = String(value[value.index(after: open)..<value.index(before: value.endIndex)])
		let body = inner.replacingOccurrences(of: ",", with: " ")
		var parts = body.split(separator: "/").map { String($0) }
		guard !parts.isEmpty else { return nil }
		var alpha: String?
		if parts.count > 1 {
			alpha = parts[1].trimmingCharacters(in: .whitespaces)
			parts = [parts[0]]
		}
		let components = parts[0]
			.split(whereSeparator: { $0 == " " || $0 == "\t" })
			.map(String.init)
		return (components, alpha)
	}

	private static func componentValue(_ text: String, scale: Double) -> Double? {
		if text.hasSuffix("%") {
			guard let number = Double(text.dropLast()) else { return nil }
			return number / 100
		}
		guard let number = Double(text) else { return nil }
		return number / scale
	}

	/// `nil` when an explicit alpha is present but unparsable, so the caller can
	/// keep its fallback token instead of silently rendering an opaque color.
	private static func alphaValue(_ text: String?) -> Double? {
		guard let text, !text.isEmpty else { return 1 }
		if text.hasSuffix("%") {
			guard let percentage = Double(text.dropLast()) else { return nil }
			return percentage / 100
		}
		return Double(text)
	}

	private static func fromFunctional(_ value: String, prefix: String) -> RGBAColor? {
		guard let (components, alpha) = splitFunction(value, prefix: prefix), components.count >= 3 else {
			return nil
		}
		guard
			let red = componentValue(components[0], scale: 255),
			let green = componentValue(components[1], scale: 255),
			let blue = componentValue(components[2], scale: 255)
		else {
			return nil
		}
		let resolvedAlpha = alpha ?? (components.count >= 4 ? components[3] : nil)
		guard let alphaValue = alphaValue(resolvedAlpha) else { return nil }
		return RGBAColor(red: red, green: green, blue: blue, alpha: alphaValue)
	}

	private static func fromHSL(_ value: String) -> RGBAColor? {
		guard let (components, alpha) = splitFunction(value, prefix: "hsl"), components.count >= 3 else {
			return nil
		}
		guard
			let hue = Double(components[0].replacingOccurrences(of: "deg", with: "")),
			let saturation = componentValue(components[1], scale: 100),
			let lightness = componentValue(components[2], scale: 100)
		else {
			return nil
		}
		guard
			let alphaValue = alphaValue(alpha ?? (components.count >= 4 ? components[3] : nil))
		else {
			return nil
		}
		return fromHSLComponents(
			hue: hue,
			saturation: saturation,
			lightness: lightness,
			alpha: alphaValue
		)
	}

	public static func fromHSLComponents(hue: Double, saturation: Double, lightness: Double, alpha: Double = 1) -> RGBAColor {
		let hueDegrees = hue.truncatingRemainder(dividingBy: 360)
		let normalizedHue = hueDegrees < 0 ? hueDegrees + 360 : hueDegrees
		let chroma = (1 - abs(2 * lightness - 1)) * saturation
		let secondary = chroma * (1 - abs((normalizedHue / 60).truncatingRemainder(dividingBy: 2) - 1))
		let match = lightness - chroma / 2

		var rgb: (Double, Double, Double)
		switch normalizedHue {
		case 0..<60: rgb = (chroma, secondary, 0)
		case 60..<120: rgb = (secondary, chroma, 0)
		case 120..<180: rgb = (0, chroma, secondary)
		case 180..<240: rgb = (0, secondary, chroma)
		case 240..<300: rgb = (secondary, 0, chroma)
		default: rgb = (chroma, 0, secondary)
		}

		return RGBAColor(
			red: rgb.0 + match,
			green: rgb.1 + match,
			blue: rgb.2 + match,
			alpha: alpha
		)
	}

	/// `oklch(0.205 0 0)` / `oklch(62.3% 0.214 259.815)` / `oklch(1 0 0 / 50%)`.
	private static func fromOKLCH(_ value: String) -> RGBAColor? {
		guard !value.contains("from ") else { return nil }
		guard let (components, alpha) = splitFunction(value, prefix: "oklch"), components.count >= 3 else {
			return nil
		}
		guard let lightness = componentValue(components[0], scale: 1) else { return nil }
		guard let chroma = Double(components[1]) else { return nil }
		guard let alphaValue = alphaValue(alpha) else { return nil }
		let hueText = components[2].replacingOccurrences(of: "deg", with: "")
		let hue: Double
		if hueText == "none" {
			hue = 0
		} else if let parsedHue = Double(hueText) {
			hue = parsedHue
		} else {
			// An unparsable hue would silently become red; let the caller fall back.
			return nil
		}
		return fromOKLCHComponents(
			lightness: lightness,
			chroma: chroma,
			hue: hue,
			alpha: alphaValue
		)
	}

	private static func fromOKLab(_ value: String) -> RGBAColor? {
		guard let (components, alpha) = splitFunction(value, prefix: "oklab"), components.count >= 3 else {
			return nil
		}
		guard
			let lightness = componentValue(components[0], scale: 1),
			let a = Double(components[1]),
			let b = Double(components[2]),
			let alphaValue = alphaValue(alpha)
		else {
			return nil
		}
		return fromOKLabComponents(lightness: lightness, a: a, b: b, alpha: alphaValue)
	}

	/// OKLCH → OKLab → linear sRGB → sRGB (Björn Ottosson's matrices).
	public static func fromOKLCHComponents(
		lightness: Double,
		chroma: Double,
		hue: Double,
		alpha: Double = 1
	) -> RGBAColor {
		let hueRadians = hue * Double.pi / 180
		return fromOKLabComponents(
			lightness: lightness,
			a: chroma * cos(hueRadians),
			b: chroma * sin(hueRadians),
			alpha: alpha
		)
	}

	public static func fromOKLabComponents(
		lightness: Double,
		a: Double,
		b: Double,
		alpha: Double = 1
	) -> RGBAColor {
		let long = pow(lightness + 0.3963377774 * a + 0.2158037573 * b, 3)
		let medium = pow(lightness - 0.1055613458 * a - 0.0638541728 * b, 3)
		let short = pow(lightness - 0.0894841775 * a - 1.2914855480 * b, 3)

		let linearRed = 4.0767416621 * long - 3.3077115913 * medium + 0.2309699292 * short
		let linearGreen = -1.2684380046 * long + 2.6097574011 * medium - 0.3413193965 * short
		let linearBlue = -0.0041960863 * long - 0.7034186147 * medium + 1.7076147010 * short

		func gamma(_ value: Double) -> Double {
			value <= 0.0031308 ? 12.92 * value : 1.055 * pow(max(value, 0), 1 / 2.4) - 0.055
		}

		return RGBAColor(
			red: gamma(linearRed),
			green: gamma(linearGreen),
			blue: gamma(linearBlue),
			alpha: alpha
		)
	}
}
